### PR TITLE
[Core] Cache job_configs instead of ray_namespace.

### PR DESCRIPTION
**Pull request by jovany-wang**
## Why are these changes needed?
We need to get not only ray_namespace config of a job. In this PR, we cache the job_configs instead of ray_namespaces, so that we can use it for other PR(For example, this PR #21249 needs the `num_java_worker_pre_process` item).

Also, before this PR, `ray_namespaces_` cache will not be cleared, and we clear the cache in this PR.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/gcs/gcs_server/gcs_job_manager.cc [hunk 1]
```diff
@@ -123,6 +124,8 @@ void GcsJobManager::ClearJobInfos(const JobID &job_id) {
   for (auto &listener : job_finished_listeners_) {
     listener(std::make_shared<JobID>(job_id));
   }
+  // Clear cache.
+  RAY_UNUSED(cached_job_configs_.erase(job_id));
```
Comment (jovany-wang) [inline src/ray/gcs/gcs_server/gcs_job_manager.cc:128]: Hum, this clearing makes `test_placement_group_2.py::test_automatic_cleanup_detached_actors` failed. The reason is PG will use `get_ray_namespace(job_id1)` after job_id1 finished.
@wuisawesome @clay4444 CC 
The stack is:
```shell
[2021-12-28 17:56:18,501 I 63827 7604794] gcs_actor_manager.cc:990: Actor created successfully, actor id = 60ff9e9b3a4224e2b337e2b602000000, job id = 02000000
[2021-12-28 17:56:18,595 C 63827 7604794] gcs_job_manager.cc:176:  Check failed: it != cached_job_configs_.end() Couldn't find job with id: 02000000
*** StackTrace Information ***
    ray::GetCallTrace()
    ray::SpdLogMessage::Flush()
    ray::SpdLogMessage::~SpdLogMessage()
    ray::RayLog::~RayLog()
    ray::gcs::GcsJobManager::GetRayNamespace()
    std::__1::__function::__func<>::operator()()
    ray::gcs::GcsPlacementGroupManager::HandleCreatePlacementGroup()
    ray::rpc::ServerCallImpl<>::HandleRequestImpl()
    boost::asio::detail::completion_handler<>::do_complete()
    boost::asio::detail::scheduler::do_run_one()
    boost::asio::detail::scheduler::run()
    boost::asio::io_context::run()
    main
    start
    0x0
```

Comment (jovany-wang) [reply src/ray/gcs/gcs_server/gcs_job_manager.cc:128]: If we don't clear it, it will waste memory of GCS, which maybe an issue in large scale cluster.

**Comment by jovany-wang** [comment]
@rkooo567 Who should I assign this to?

**Comment by jovany-wang** [comment]
Refactoring...

File: src/ray/gcs/gcs_server/gcs_job_manager.cc [hunk 2]
```diff
@@ -55,9 +53,8 @@ void GcsJobManager::HandleAddJob(const rpc::AddJobRequest &request,
       }
       RAY_LOG(INFO) << "Finished adding job, job id = " << job_id
                     << ", driver pid = " << mutable_job_table_data.driver_pid();
-      ray_namespaces_[job_id] = mutable_job_table_data.config().ray_namespace();
-      cache_num_java_worker_per_processes_[job_id] =
-          mutable_job_table_data.config().num_java_workers_per_process();
+      cached_job_configs_[job_id] =
+          std::make_unique<rpc::JobConfig>(mutable_job_table_data.config());
```
Comment (jovany-wang) [inline src/ray/gcs/gcs_server/gcs_job_manager.cc:57]:
```suggestion
          std::make_shared<rpc::JobConfig>(mutable_job_table_data.config());
```

File: src/ray/gcs/gcs_server/gcs_job_manager.cc [hunk 3]
```diff
@@ -23,10 +23,8 @@ void GcsJobManager::Initialize(const GcsInitData &gcs_init_data) {
   for (auto &pair : gcs_init_data.Jobs()) {
     const auto &job_id = pair.first;
     const auto &job_table_data = pair.second;
-    const auto &ray_namespace = job_table_data.config().ray_namespace();
-    ray_namespaces_[job_id] = ray_namespace;
-    cache_num_java_worker_per_processes_[job_id] =
-        job_table_data.config().num_java_workers_per_process();
+    cached_job_configs_[job_id] =
+        std::make_unique<rpc::JobConfig>(job_table_data.config());
```
Comment (jovany-wang) [inline src/ray/gcs/gcs_server/gcs_job_manager.cc:27]:
```suggestion
        std::make_shared<rpc::JobConfig>(job_table_data.config());
```

**Comment by jovany-wang** [comment]
@WangTaoTheTonic CC

**Comment by jovany-wang** [comment]
This is ready for review.

**Review by rkooo567 — COMMENTED** [review]
LGTM. Can you add a test case to the job manager?

**Comment by jovany-wang** [comment]
> LGTM. Can you add a test case to the job manager?

Sounds good. Will do.

**Comment by jovany-wang** [comment]
> LGTM. Can you add a test case to the job manager?

gcs_job_manager_test was added.

File: src/ray/gcs/gcs_server/gcs_actor_manager.h [hunk 4]
```diff
@@ -529,12 +528,10 @@ class GcsActorManager : public rpc::ActorInfoHandler {
   /// This method MUST BE IDEMPOTENT because it can be called multiple times during
   /// actor destroy process.
   std::function<void(const ActorID &)> destroy_owned_placement_group_if_needed_;
-  /// A callback to get the namespace an actor belongs to based on its job id. This is
+  /// A callback to get the job config of an actor belongs to based on its job id. This is
```
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_actor_manager.h:531]:
```suggestion
  /// A callback to get the job config of an actor based on its job id. This is
```

File: src/ray/gcs/gcs_server/gcs_job_manager.h [hunk 5]
```diff
@@ -68,11 +66,8 @@ class GcsJobManager : public rpc::JobInfoHandler {
   /// Listeners which monitors the finish of jobs.
   std::vector<std::function<void(std::shared_ptr<JobID>)>> job_finished_listeners_;
 
-  /// A cached mapping from job id to namespace.
-  std::unordered_map<JobID, std::string> ray_namespaces_;
-
-  /// A cached mapping from job id to num_java_workers_per_process.
-  std::unordered_map<JobID, int32_t> cache_num_java_worker_per_processes_;
+  /// A cached mapping from job id to job config.
+  absl::flat_hash_map<JobID, std::shared_ptr<rpc::JobConfig>> cached_job_configs_;
```
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_job_manager.h:70]: Nice cleanup!

**Review by ericl — APPROVED** [review]

**Pull request merged**